utils/node-loader/src/batch_pool.rs

```diff
@@ -45,329 +47,206 @@ impl<Rng: LoaderRng> BatchPool<Rng> {
-                        logs.len() + 1
-                    ));
-                }
-            }
-
-            Ok(Report {
-                logs,
-                codes: BTreeSet::new(),
-                program_ids: BTreeSet::new(),
-                blocks_stopped,
-            })
+            let (ex_results, block_hash) = api.send_message_batch(args).await?;
+            let messages = process_ex_results(ex_results);
+            process_events(api.into_gear_api(), messages, block_hash, false).await
         }
         Batch::CreateProgram(args) => {
-            let args = args.into_iter().map(|v| v.into());
-
-            let (ex_results, batch_block_hash) = api.create_program_bytes_batch(args).await?;
-
-            let mut init_messages = BTreeMap::new();
-
-            for r in ex_results {
-                match r {
-                    Ok((mid, pid)) => {
-                        init_messages.insert(mid, pid);
-                    }
-                    Err(e) => logs.push(format!(
-                        "[#{:<2}] Extrinsic failure: '{:?}'",
-                        logs.len() + 1,
-                        e
-                    )),
-                }
+            let (ex_results, block_hash) = api.create_program_batch(args).await?;
+            let messages = process_ex_results(ex_results);
+            process_events(api.into_gear_api(), messages, block_hash, true).await
+        }
+    }
+}
+
+fn process_ex_results<Key: Ord, Value>(
+    ex_results: impl IntoIterator<Item = GClientResult<(Key, Value)>>,
+) -> BTreeMap<Key, (Value, CallId)> {
+    let mut res = BTreeMap::<Key, (Value, CallId)>::new();
+
+    for (i, r) in ex_results.into_iter().enumerate() {
+        let call_id = i + 1;
+        match r {
+            Ok((key, value)) => {
+                res.insert(key, (value, call_id));
+                tracing::debug!("[Call with id: {call_id}]: Successfully executed.")
             }
+            Err(e) => tracing::debug!("[Call with id: {call_id}]: Failed: '{e:?}'"),
+        }
+    }
 
-            let results: Result<Vec<(MessageId, Option<String>)>>;
+    res
+}
 
-            let now = utils::now();
+async fn process_events(
+    api: GearApi,
+    mut messages: BTreeMap<MessageId, (ProgramId, usize)>,
+    block_hash: H256,
+    collect_programs: bool,
+) -> Result<Report> {
+    let results: GClientResult<Vec<(MessageId, Option<String>)>>;
+    let now = utils::now();
+    // States what amount of blocks we should wait for taking all the events about successful `messages` execution
+    let wait_for_events_blocks = 10;
+    // Multiply on five to be 100% sure if no events occurred, than node is crashed
+    let wait_for_events_millisec = api.expected_block_time()? as usize * wait_for_events_blocks * 5;
+
+    loop {
+        let r = match api.events_since(block_hash, wait_for_events_blocks).await {
+            Ok(mut v) => v.err_or_succeed_batch(messages.keys().copied()).await,
```

Comment:
Calling `messages.keys` in the loop seems not good unless `messages` changes inside the loop.

Reply:
discussed